subscriber.go

```diff
@@ -778,15 +793,77 @@ func (h *handler) handleAsync(ctx context.Context, nextCid cid.Cid, syncer Synce
+
+type (
+	// SegmentSyncActions allows the user to control the flow of segmented sync by either choosing
+	// which CID should be synced in the next sync cycle or setting the error that should mark the
+	// sync as failed.
+	SegmentSyncActions interface {
+		// SetNextSyncCid sets the cid that will be synced in the next segmented sync. Note that
+		// the last call to this function during a segmented sync cycle dictates which CID will be
+		// synced in the next cycle.
+		//
+		// At least one call to this function must be made for the segmented sync cycles to continue.
+		// Because, otherwise the CID that should be used in the next segmented sync cycle cannot
+		// be known.
+		//
+		// If no calls are made to this function or next CID is set to cid.Undef, the sync will
+		// terminate and any CIDs that are synced so far will be included in a SyncFinished event.
+		SetNextSyncCid(cid.Cid)
+
+		// FailSync fails the sync and returns the given error as soon as the current segment sync
+		// finishes. Note that the last call to this function during a segmented sync cycle takes
+		// dictates the error value.
+		// Passing nil as error will cancel sync failure.
+		FailSync(error)
+	}
+	// SegmentBlockHookFunc is called for each synced block, similarly to BlockHookFunc. Except that
+	// it provides SegmentSyncActions to the hook allowing the user to contl the flow of segmented
+	// sync by determining which CID should be used in the next segmented sync cycle by decoding the
+	// synced block.
+	// SegmentSyncActions also allows the user to signal any errors that may occur during the hook
+	// execution to terminate the sync and mark it as failed.
+	SegmentBlockHookFunc func(peer.ID, cid.Cid, SegmentSyncActions)
+	segmentedSync        struct {
+		nextSyncCid *cid.Cid
+		err         error
+	}
+)
+
+func (ss *segmentedSync) SetNextSyncCid(c cid.Cid) {
+	ss.nextSyncCid = &c
+}
+
+func (ss *segmentedSync) FailSync(err error) {
+	ss.err = err
+}
+
+func (ss *segmentedSync) reset() {
+	ss.nextSyncCid = nil
+	ss.err = nil
+}
+
 // handle processes a message from the peer that the handler is responsible for.
-func (h *handler) handle(ctx context.Context, nextCid cid.Cid, sel ipld.Node, wrapSel bool, syncer Syncer, hook BlockHookFunc) ([]cid.Cid, error) {
+func (h *handler) handle(ctx context.Context, nextCid cid.Cid, sel ipld.Node, wrapSel bool, syncer Syncer, bh BlockHookFunc, segdl int64, segbh SegmentBlockHookFunc) ([]cid.Cid, error) {
 	h.syncMutex.Lock()
 	defer h.syncMutex.Unlock()
 	log := log.With("cid", nextCid, "peer", h.peerID)
 
-	// This is not set to nil so we can get a pointer.
-	syncedCids := []cid.Cid{}
-	hook = WrapBlockHookWithSyncedCidTracker(&syncedCids, hook)
+	segSync := &segmentedSync{
+		nextSyncCid: &nextCid,
+	}
+
+	var syncedCids []cid.Cid
+	hook := func(p peer.ID, c cid.Cid) {
+		syncedCids = append(syncedCids, c)
+		if segbh != nil {
+			segbh(p, c, segSync)
+		}
+		if bh != nil {
+			bh(p, c)
+		}
```

Comment:
Does the calling order matter here?

Reply:
Not sure why anyone would provide both functions, so was not sure if there was any implicit purpose for one or the other.